src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Unix.cs

```diff
@@ -869,6 +1183,59 @@ public void HandleEvent(SocketAsyncContext context)
                 op.Dispatch();
             }
 
+            internal void BatchOrDispatch(SocketAsyncContext context, Span<Interop.Sys.IoControlBlock> ioControlBlocks, Span<AsyncOperation> batchedOperations, ref int batchedCount)
+            {
+                AsyncOperation nextOperation;
+                using (Lock())
+                {
+                    switch (_state)
+                    {
+                        case QueueState.Ready:
+                            Debug.Assert(_tail == null, "State == Ready but queue is not empty!");
+                            _sequenceNumber++;
+                            return;
+
+                        case QueueState.Waiting:
+                            Debug.Assert(_tail != null, "State == Waiting but queue is empty!");
+                            _state = QueueState.Processing;
+                            nextOperation = _tail.Next;
+
+                            while (batchedCount < ioControlBlocks.Length && nextOperation.TryBatch(context, batchedCount, ref ioControlBlocks[batchedCount]))
+                            {
+                                batchedOperations[batchedCount++] = nextOperation;
+                                _batchedOperationsCount++;
+
+                                if (nextOperation == nextOperation.Next)
+                                {
+                                    // we have batched some operations and reached the end of the queue
+                                    return;
+                                }
+
+                                nextOperation = nextOperation.Next;
+                            }
+                            break;
+
+                        case QueueState.Processing:
+                            Debug.Assert(_tail != null, "State == Processing but queue is empty!");
+                            _sequenceNumber++;
+                            return;
+
+                        case QueueState.Stopped:
+                            Debug.Assert(_tail == null);
+                            return;
+
+                        default:
+                            Environment.FailFast("unexpected queue state");
+                            return;
+                    }
+
+                    // we can't batch more (batchedCount < ioControlBlocks.Length)
+                    // or nextOperation.TryAsBatch failed
+                    // so we dispatch outside of lock
+                    nextOperation?.Dispatch();
```

Comment:
important: we are performing the actual dispatch outside the lock

Reply:
@tmds this is where the operation is dispatched if  we are unable to batch it

Reply:
If we can get from `HandleBatchEvent` to this point, then things are good.
Otherwise an operation that got added to the queue while a batched operation was executing, will sit idle in the queue waiting for an event that may never come.

Reply:
We get here before `HandleBatchEvent`.

When we receive an epoll notification we call `AddWaitingOperationsToBatch` which internally calls `queue.BatchOrDispatch` (the commented method). This method either adds operations to the batch or is dispatching them.

Later `HandleBatchEvent` is handling only what was added to the batch and sent to AIO.